### DISCUSSION TITLE
Mojo 🔥 available for local download!

### DISCUSSION THREAD

**Discussion by goldiegadde** [modular/modular#568] [Announcements]
Hey everyone 👋,

We are very excited to [announce that Mojo](https://www.modular.com/blog/mojo-its-finally-here)🔥 is now available for [download](https://developer.modular.com/signup) locally to your machine! ❤️‍🔥🚀

Beyond a compiler, the Mojo SDK includes **a full set of developer and IDE tools 🛠 that make it easy to build and iterate on Mojo applications.** It is available on Linux/X86 today and we will be adding Mac and Windows support (and many other features) in the weeks ahead.

Get started now on [www.modular.com](http://www.modular.com), download our [VS code extension](https://marketplace.visualstudio.com/items?itemName=modular-mojotools.vscode-mojo), and see our [Mojo examples in Github](https://github.com/modularml/mojo/tree/main/examples) and [changelog](https://docs.modular.com/mojo/changelog.html#v0.2.1-2023-09-07) for additional details! 

**Let’s build the future together!🔥**

**Comment by dom96** [comment]
I'm sure I'm not the only one but this is what I've been waiting for since Mojo was announced :)

**Comment by azkadev** [comment]
I tried to install Mojo but why is it taking so long? Is it because my internet speed is 20 Mbps 

![Screenshot from 2023-09-08 01-19-28](https://github.com/modularml/mojo/assets/82513502/2b3da3f2-4fd8-4ae1-81e7-820b9cc01261)


/ this website is down?
https://packages.modular.com/mojo
![Screenshot from 2023-09-08 01-21-43](https://github.com/modularml/mojo/assets/82513502/e2344ff0-01fb-49a8-869b-8f994b1ea1a7)

**Reply by pkeorley** [reply]
https://docs.modular.com/mojo/roadmap.html#mojo-sdk-known-issues
![image](https://github.com/modularml/mojo/assets/59422798/b3925c82-5276-4d61-a0d8-614fac995257)

**Reply by systra72** [reply]
Has it hanged or something if it look like this after 10 minutes:

Installing collected packages: textwrap3, find_libpython, tqdm, entrypoints, click, ansiwrap, papermill
Successfully installed ansiwrap-0.8.4 click-8.1.7 entrypoints-0.4 find_libpython-0.3.0 papermill-2.4.0 textwrap3-0.9.2 tqdm-4.66.1
Testing `MODULAR_HOME=/home/codespace/.modular`
* `/home/codespace/.modular/pkg/packages.modular.com_mojo/bin/mojo`...
TEST: `mojo --help`... OK
TEST: `mojo run --help`... OK
TEST: `mojo build test_mandelbrot.mojo`... OK
TEST: `mojo build test_python.mojo`... OK
TEST: `mojo demangle`... OK
TEST: `mojo format`... OK
TEST: `mojo package`... OK
TEST: `mojo test_mandelbrot.mojo`... OK
TEST: `mojo test_python.mojo`... OK

**Reply by garam-kim1** [reply]
Adding more cpus solves this (https://github.com/modularml/mojo/issues/538)

**Reply by rarebreed** [reply]
I am having that issue on my chromebook while running the linux container.  I chalked it up to it being a debian rather than ubuntu system.  That being said, I am running mojo fine so far on a linux mint 21.2 system

**Reply by megamorf** [reply]
> Adding more cpus solves this (https://github.com/modularml/mojo/issues/538)

That did not fix the issue for me (bumped number of cores from 2 to 4, still stuck after the test_python.mojo step).

**Reply by drunkwcodes** [reply]
@megamorf Did you check the wsl version by `wsl -l -v`? It should be 2.
Please follow #538 for the details.

**Reply by megamorf** [reply]
@drunkwcodes My setup doesn't use Windows and details about it have already been posted in that issue, but thanks 😉

**Comment by martin-braun** [comment]
Great news and good job. Is there any estimate on native support on Macs? Also I'd like to know if there are any road blocks for not choosing Debian support?

I can't really leverage Mojo a lot from a VM when working with the local file system, but I still want to express my gratitude for this major step.

**Comment by love01211** [comment]
I'm using Windows Subsystem for Linux (WSL) version Linux LAPTOP-MK90AVHL 5.15.90.1-microsoft-standard-WSL2 #1 SMP Fri Jan 27 02:56:13 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux. When I try to install something, I receive the error message 'Failed to update via apt-get update.' However, I can successfully curl google.com. Here is the error message I'm getting.

![image](https://github.com/modularml/mojo/assets/31766073/82f231a2-f33f-4f3c-b713-44c30d16f42c)

**Reply by love01211** [reply]
The problem has been resolved, and I am now able to use Clash for VPN connections within the Windows Subsystem.

**Comment by KatieUmbra** [comment]
Any plans on creating developing an lsp?

**Reply by jackos** [reply]
The Mojo SDK comes with an LSP and this extension uses it: https://marketplace.visualstudio.com/items?itemName=modular-mojotools.vscode-mojo

**Reply by KatieUmbra** [reply]
just realized, that makes sense, sorry for the question I should've phrased it better

**Comment by ankur-gupta** [comment]
I am excited about Mojo but putting the download behind an "email"-wall is quite the disappointing start. According to the [privacy policy](https://www.modular.com/legal/privacy), email is shared with **"Advertising Partners, Analytics Partners"**. This will hamper adoption.

**Reply by rarebreed** [reply]
It's stated that eventually mojo [will be open sourced]( https://docs.modular.com/mojo/faq.html#will-mojo-be-open-sourced), so I assume that would eventually bypass the need for an email sign up wall.  Not sure when, but my feeling is that what they have now is still pretty far from a github open source release (glad to be wrong though!).  However, the following FAQs may be useful:

- https://docs.modular.com/mojo/faq.html#why-not-develop-mojo-in-the-open-from-the-beginning
- https://docs.modular.com/mojo/faq.html#does-the-mojo-sdk-collect-telemetry

Just from my initial playing around with mojo, the standard library is still in a very very early stage.  It's basically meant for pure computation.  Unless you interop with python, you don't get true IO (other than print) and not even a dictionary type yet.  That means you can't do pure mojo network operations or reading in of files yet.

Also, it's still not fully interoperable with python (or I haven't seen a good way to do this from the documentation yet).  For example, mojo [doesn't yet support keyword arguments](https://docs.modular.com/mojo/roadmap.html#small-independent-features), nor does it support unpacking (eg **kwargs).  This makes calling python functions that require keyword type arguments (eg, the * operator is in the parameter declaration), not doable from mojo.

I'd say for now, mojo will really be of interest to AI and pure HPC devs where only pure computation is needed.  I'd wait a bit for mojo to become more usable as a [general purpose language](https://docs.modular.com/mojo/faq.html#is-mojo-only-for-ai-or-can-it-be-used-for-other-stuff). I'd love for it to become open sourced soon so I could pitch in and help in some way.  But I agree with their viewpoint, that sometimes you need a tight vision to get things done, especially at this early stage.

**Reply by ankur-gupta** [reply]
I agree with most of it. 

Yes, mojo is very early stage language and they need to retain a lot more control over design decisions. Removing the email-wall wouldn't take away their control over design decisions. They also don't have to put the source code on GitHub. They can still remove the email-wall without having to put the source code on GitHub. 

In my opinion: Forcing people to provide their email addresses (which evidently can be used for advertising) to try out such an early-stage language will only hamper its adoption. Poor adoption is bad for everyone not just the creators of mojo. Many languages sputter out because of poor adoption.

**Reply by rarebreed** [reply]
I agree that it would be nice if the email was opt-in.  I for one don't mind though, but I can see how this will make others pause.  They even say it's better to use your work email address, so that means they would prefer to target users who could get their companies to try out mojo/Modular Engine.  

At this early stage, I can't begrudge Modular for wanting or needing some insight into who is interested in Mojo (and by extension, interested in accelerating ML models and therefore their Modular Engine product where they make money).  I remember when Mozilla ran into financial difficulties, a lot of folks were concerned about the viability of the Rust language.  Fortunately, rust was already being used by big companies like Amazon, Google and Microsoft, so thankfully, corporate sponsorship helped form the Rust Foundation.  Honestly, that's one of my fears for mojo: having enough capital to build out mojo to the point that it is adopted by other Big Tech.  Mozilla was (is) an established company that could afford to bankroll rust core developers in its early stages and not a venture capital start up needing angel investors.

I could be wrong, but I think the biggest early adopters are going to be individuals curious about what mojo can do, and less so companies doing investigation into it.  Modular needs to turn the individual early adopters into evangelists to their employers.  Rust only got as far as it did by having a small but extremely dedicated following pushing it (the Most Loved Language on Stack Overflow surveys helped I am sure).  So I think the biggest thing Modular can do to help with adoption is:

- Make the learning curve/transition as easy as possible (eg like [rust's free book](https://doc.rust-lang.org/book/) would be awesome)
- A list of all the things mojo still needs to be a superset of python (the roadmap and rough edges helps, but a tick box would be nice)
    - I don't know if python has a compatibility test suite for alternate python implementations, but if it does, show where mojo fails
    - If one doesn't exist, maybe [rust python's test suite](https://rustpython.github.io/pages/regression-tests-results.html) could be leveraged
- Prioritize more python interop working as soon as possible to make up for the missing mojo standard library
- Make development as nice or better than Rust (eg protocols/traits, Algebraic Data Types, easier to understand lifetimes, etc)

I understand Modular is prioritizing mojo working for their use cases, but I think if they do the above, it will help adoption.

**Comment by thinkerou** [comment]
when release mac version? thanks!

**Comment by cataYT** [comment]
I downloaded the vs code extension but the auto complete and other features doesnt seem to work.

**Comment by JohnDinsBBR** [comment]
The Linux kernal is disabled on my machine, when will a native windows version be available?

**Reply by astearon** [reply]
Same issue here! 1 year later and still no change... Native Windows version when?

**Comment by Swarnodip-Nag** [comment]
happily running mojo but speed is not that get promised

**Reply by soraros** [reply]
Hey @Swarno-Coder, do you have a runnable example demonstrating the said slowdown you are experiencing? That can maybe helpful to the Mojo team.

**Comment by nevakrien** [comment]
![Screenshot from 2024-02-10 18-30-18](https://github.com/modularml/mojo/assets/101988414/921980e4-ebbf-4c0a-a44a-201c20813629)
I ran it on ubuntu and it just dosent boot a new project... setup tools broke first so I installed it with build and then it broke compltly. I can still import in python tho


(base) user@user-System-Product-Name:~/repo_packages/Mojo-0.1$ conda activate mojo
(mojo) user@user-System-Product-Name:~/repo_packages/Mojo-0.1$ which mojo manager
(mojo) user@user-System-Product-Name:~/repo_packages/Mojo-0.1$ which mojo_manager
/home/user/anaconda3/envs/mojo/bin/mojo_manager
(mojo) user@user-System-Product-Name:~/repo_packages/Mojo-0.1$ mojo_manager
  File "/home/user/anaconda3/envs/mojo/bin/mojo_manager", line 35
    print 'Creating new Mojo project: %s...' % project_name
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
SyntaxError: Missing parentheses in call to 'print'. Did you mean print(...)?
(mojo) user@user-System-Product-Name:~/repo_packages/Mojo-0.1$ mojo_manager -p test_project
  File "/home/user/anaconda3/envs/mojo/bin/mojo_manager", line 35
    print 'Creating new Mojo project: %s...' % project_name
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
SyntaxError: Missing parentheses in call to 'print'. Did you mean print(...)?
(mojo) user@user-System-Product-Name:~/repo_packages/Mojo-0.1$
